crates/sui-swarm/src/memory/node.rs

```diff
@@ -0,0 +1,228 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use futures::FutureExt;
+use std::thread;
+use sui_config::NodeConfig;
+use sui_node::SuiNode;
+use sui_types::base_types::SuiAddress;
+use tap::TapFallible;
+use tracing::error;
+
+/// A handle to an in-memory Sui Node.
+///
+/// Each Node is attempted to run in isolation from each other by running them in their own tokio
+/// runtime in a separate thread. By doing this we can ensure that all asynchronous tasks
+/// associated with a Node are able to be stopped when desired (either when a Node is dropped or
+/// explicitly stopped by calling [`Node::stop`]) by simply dropping that Node's runtime.
+#[derive(Debug)]
+pub struct Node {
+    thread: Option<Container>,
+    config: NodeConfig,
+    runtime_type: RuntimeType,
+}
+
+impl Node {
+    /// Create a new Node from the provided `NodeConfig`.
+    ///
+    /// The Node is returned without being started. See [`Node::spawn`] or [`Node::start`] for how to
+    /// start the node.
+    ///
+    /// [`NodeConfig`]: sui_config::NodeConfig
+    pub fn new(config: NodeConfig) -> Self {
+        Self {
+            thread: None,
+            config,
+            runtime_type: RuntimeType::SingleThreaded,
+        }
+    }
+
+    /// Return the `name` of this Node
+    pub fn name(&self) -> SuiAddress {
+        self.config.sui_address()
+    }
+
+    /// Start this Node, returning a handle that will resolve when the node has completed starting
+    /// up.
+    pub fn spawn(&mut self) -> Result<tokio::sync::oneshot::Receiver<()>> {
+        let (startup_reciever, node_handle) =
+            Container::spawn(self.config.clone(), self.runtime_type);
+        self.thread = Some(node_handle);
+        Ok(startup_reciever)
+    }
+
+    /// Start this Node, waiting until its completely started up.
+    pub async fn start(&mut self) -> Result<()> {
+        let startup_reciever = self.spawn()?;
+        startup_reciever.await?;
+        Ok(())
+    }
+
+    /// Stop this Node
+    pub fn stop(&mut self) {
+        self.thread = None;
+    }
+
+    /// Perform a health check on this Node by:
+    /// * Checking that the node is running
+    /// * Calling the Node's gRPC Health service
+    pub async fn health_check(&self) -> Result<(), HealthCheckError> {
+        let thread = self.thread.as_ref().ok_or(HealthCheckError::NotRunning)?;
+        if !thread.is_alive() {
+            return Err(HealthCheckError::NotRunning);
+        }
+
+        let channel = mysten_network::client::connect(self.config.network_address())
```

Comment:
Does this run under a timeout? ([see also](https://github.com/MystenLabs/narwhal/issues/196))

Reply:
I believe there's a default although a more aggressive one can be configured.